### DISCUSSION TITLE
DNS sensibility

### DISCUSSION THREAD

**Discussion by PaPaTheGMan** [Ylianst/MeshCentral#5137] [Q&A]
I've found that if a client switches from primary DNS to an alternate, MeshCentral is unable to find/attach to the agent.  Any ideas as to why this is?  It's a bit of a pain to SSH into the client separately to reassign it's DNS server before MeshCentral will find it.

Thanks.

**Comment by SomeGuru** [comment]
Are you stating that as an Agent is installed on say dns 8.8.8.8 and then you decide to change the agent machine to say dns 1.0.0.1, that it doesn't come online or are you saying that when you go from an internal say 10.21.32.X. That dns of 10.11.22.33 changed to 10.22.33.44 it doesn't connect after that? Trying to understand your plight here..

-SomeGuru

**Comment by PaPaTheGMan** [comment]
Hi, thanks for reaching out.  It's actually a bit of both.  I use Pi-Hole on my network.  My router is the DHCP server with DNS servers defined as 192.168.x.x (Pi-Hole) and 1.1.1.1 (It's not Cloudflare that's the issue, I've tried other public DNS servers and the same situation happens.).  I've found that I need a backup Outside/public DNS server for when the Pi-Hole server decides to take a nap.  The MeshCentral server is a node on the 192.168.x.x network.  Anyway, my home-lab servers with MeshCentral's agent installed are all headless and will normally be using 192.168.xx.  MeshCentral can find them just fine.  But , at least once a day, the MeshCentral server can't find the agent to one or more of them.  When I inspect them I see that their DNS assignment has changed to 1.1.1.1.  When I re-set the DNS to 192.168.x.x, they connect back to Mesh-Central just fine.

For example, as I type this, of the 15 agents I have installed, MeshCentral only recognizes 13.  The two which are unavailable all show their current DNS to be 1.1.1.1.  When I force a change to 192.168.x.x, MeshCentral finds them again.

When I'm home this issue isn't all that critical, just annoying.  But when I'm away, then I'm blocked out of access. I want to use MeshCentral as a KVM for both terminal and desktop access to my Home-Lab servers.

What I don't understand is why MeshCentral is so sensitive to the DNS server.  Don't most people have a primary and backup DNS?

Thanks.

**Reply by si458** [reply]
What's the dns name u are using?
If it's an internal one, `meshcentral.local` then cloudflare won't be able to resolve it which is why it's failing...
Also does the dns name point to the 192.168 ip or ur external ip?

**Comment by PaPaTheGMan** [comment]
I don't use a name, just the IP addresses - 192.168.x.x, 1.1.1.1.  I'm not sure what you are asking in the 2nd question.  The only place I put in any DNS setting is my router which has 192.168.x.x, 1.1.1.1

**Reply by SomeGuru** [reply]
What is your config.json looking like with domain redacted of course... interested to see how you are configured there...

**Comment by PaPaTheGMan** [comment]
Config.json attached as is a copy of the docker create file used.

**Reply by PaPaTheGMan** [reply]
Any insights?  This keeps happening and is getting rather annoying.  Should I just put this in as a bug?

**Reply by PrplHaz4** [reply]
I don't think you've answered the questions above yet.  Is your meshcentral behind a publicly available domain name? IOW, can Cloudflare (1.1.1.1) resolve the DNS of your meshcentral server?

If it's not publicly-resolveable, it can only be found when you're using local DNS (192.x) - that would explain the behavior you observed.

**Comment by PaPaTheGMan** [comment]
Sorry, I thought I answered.  I'm not talking about a public server at all.  I'm talking about my MeshCentral server being on the same LAN as the other servers.  It is reachable within my network by its 192.168.x.x address.  Is it a requirement that MeshCentral have a FQDN?

**Reply by PrplHaz4** [reply]
No it's not a requirement - but I saw the HOSTNAME variable in your docker run.  You may be able to prove it's a DNS issue by removing 1.1.1.1 as an option and see if that fixes the issue - if so it likely means that somewhere is relying on internal name resolution.

**Comment by PaPaTheGMan** [comment]
I just tried what you suggested.  Once the network refresh completed, all of MeshCentral's clients went offline.  Resetting the DNS to my local server, brought the clients back.

What I had done in the past on a couple of servers was to manually switch their local DNS around.  MeshCentral loses contact with a client on a server that has 1.1.1.1 as its DNS.  It then regains contact when the DNS server is switched back to my internal server.  This is what made me think MeshCentral was being too sensitive to a client's DNS setting and prompted my original query.  Also, even if the server where MeshCentral resides, switches to 1.1.1.1, it can't find its client, but can still find others that have their DNS set to the internal DNS server.

I've also tried to install the client on a server that currently points to 1.1.1.1.  MeashCentral won't install.

I don't have a FQDN assigned to any of my servers. I use a dynamic DNS facility to point to my ISP address.  Currently, MeshCentral sits behind a Nginx Proxy Manager reverse proxy.  If I ping MeshCentral's assigned NPM Domain Name from within my network, I get the IP of the NPM server.  If I ping it from outside, I get my ISP address.   My router's firewall forwards all '80 & '443 traffic to the NPM server.

I don't know enough about networking to understand why these servers switch themselves automatically to the secondary DNS.  That's a future learning opportunity for me.  I presume something in my primary DNS studders making the sever think the primary is down.

Hopefully this additional information gives you something to work with.

**Reply by PrplHaz4** [reply]
My understanding is that these are not treated as primary and secondary DNS servers - you should be ready for any request to be served by either one.

The solution I have for this is split-horizon DNS - everything internally is resolved by my gateway to the internal IP and everything external resolves via public dns to my dynamic ISP IP.

**Comment by PaPaTheGMan** [comment]
Interesting.  As I indicated, I'm not really versed in networking, so I don't really understand how a 'split-horizon DNS ' works.  Is there something I need to do differently?  Or, are the changes on your end?
